klippa/src/cblc.rs

```diff
@@ -0,0 +1,762 @@
+                }
+            }
+        }
+
+        if num_sizes == 0 || cbdt_out.is_empty() {
+            return Err(SubsetError::SubsetTableError(Cblc::TAG));
+        }
+
+        s.copy_assign(num_sizes_pos, num_sizes);
+        builder.add_raw(Cbdt::TAG, cbdt_out);
+        Ok(())
+    }
+}
+
+impl<'a> SubsetTable<'a> for BitmapSize {
+    // (Cblc table, CBDT table, src_bitmapsize_bytes, cbdt_out)
+    type ArgsForSubset = (&'a Cblc<'a>, &'a Cbdt<'a>, &'a [u8], &'a mut Vec<u8>);
+    type SubsetOutput = ();
+    fn subset(
+        &self,
+        plan: &Plan,
+        s: &mut Serializer,
+        args: Self::ArgsForSubset,
+    ) -> Result<(), SerializeErrorFlags> {
+        let cblc = args.0;
+        let cbdt = args.1;
+        let src_bytes = args.2;
+        let cbdt_out = args.3;
+
+        if self.start_glyph_index() > plan.glyphset.last().unwrap()
+            || self.end_glyph_index() < plan.glyphset.first().unwrap()
+        {
+            return Err(SerializeErrorFlags::SERIALIZE_ERROR_EMPTY);
+        }
+
+        let snap = s.snapshot();
+        let offset_pos = s.embed_bytes(src_bytes)?;
+        let Ok(index_subtable_list) = self.index_subtable_list(cblc.offset_data()) else {
+            return Err(s.set_err(SerializeErrorFlags::SERIALIZE_ERROR_READ_ERROR));
+        };
+
+        s.push()?;
+        match index_subtable_list.subset(plan, s, (cbdt, cbdt_out)) {
+            Ok((min_gid, max_gid, table_list_size, num_subtables)) => {
+                let Some(obj_idx) = s.pop_pack(true) else {
+                    return Err(s.error());
+                };
+                let _ = s.add_link(
+                    offset_pos..offset_pos + Offset32::RAW_BYTE_LEN,
+                    obj_idx,
+                    OffsetWhence::Head,
+                    0,
+                    false,
+                );
+
+                //update table list size, biye pos = 4
+                s.copy_assign(offset_pos + 4, table_list_size as u32);
+                // update number of index subtable, byte_pos = 4
+                s.copy_assign(offset_pos + 8, num_subtables as u32);
+
+                // startGlyphIndex, byte_pos = 40
+                s.copy_assign(offset_pos + 40, min_gid.to_u32() as u16);
+
+                // endGlyphIndex, byte_pos = 42
+                s.copy_assign(offset_pos + 42, max_gid.to_u32() as u16);
+            }
+            Err(SerializeErrorFlags::SERIALIZE_ERROR_EMPTY) => {
+                s.pop_discard();
+                s.revert_snapshot(snap);
+                return Err(SerializeErrorFlags::SERIALIZE_ERROR_EMPTY);
+            }
+            Err(e) => {
+                return Err(e);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<'a> SubsetTable<'a> for IndexSubtableList<'a> {
+    type ArgsForSubset = (&'a Cbdt<'a>, &'a mut Vec<u8>);
+    // min_gid(new), max_gid(new), indexSubtableListSize, numberOfIndexSubtables
+    type SubsetOutput = (GlyphId, GlyphId, usize, usize);
```

Comment:
you could consider defining a simple struct for passing in these args, so that you have named fields? in this case in particular it seems easy to get the argument order wrong.